File: public/docs/_examples/architecture/ts/app/backend.service.ts
import {Injectable, Type} from '@angular/core';
import {Logger} from './logger.service';
import {Hero} from './hero';

const HEROES = [
        new Hero('Windstorm', 'Weather mastery'),
        new Hero('Mr. Nice', 'Killing them with kindness'),
        new Hero('Magneta', 'Manipulates metalic objects')
      ];

@Injectable()
export class BackendService {
  constructor(private _logger: Logger) {}

  getAll(type:Type) : PromiseLike<any[]>{
    if (type === Hero) {
      // TODO get from the database
      return Promise.resolve<Hero[]>(HEROES);
    }
    let err = new Error('Cannot get object of this type');
    this._logger.error(err);
    throw err;
  }
}

File: public/docs/_examples/architecture/dart/lib/backend_service.dart
import 'dart:async';

import 'package:angular2/core.dart';

import 'hero.dart';
import 'logger_service.dart';

@Injectable()
class BackendService {
  static final _mockHeroes = [
    new Hero('Windstorm', 'Weather mastery'),
    new Hero('Mr. Nice', 'Killing them with kindness'),
    new Hero('Magneta', 'Manipulates metalic objects')
  ];

  final Logger _logger;

  BackendService(Logger this._logger);

  Future<List> getAll(type) {
    // TODO get from the database
    if (type == Hero) return new Future.value(_mockHeroes);

    var msg = 'Cannot get object of this type';
    _logger.error(msg);
    throw new Exception(msg);
  }
}
